packages/engine-core/src/data-proxy/DataProxyEngine.ts

```diff
@@ -219,11 +227,73 @@ export class DataProxyEngine extends Engine {
     }
   }
 
-  // TODO: figure out how to support transactions
-  transaction(): Promise<any> {
-    throw new NotImplementedYetError('Interactive transactions are not yet supported', {
-      clientVersion: this.clientVersion,
-    })
+  /**
+   * Send START, COMMIT, or ROLLBACK to the Query Engine
+   * @param action START, COMMIT, or ROLLBACK
+   * @param headers headers for tracing
+   * @param options to change the default timeouts
+   * @param info transaction information for the QE
+   */
+  async transaction(action: 'start', headers: Tx.TransactionHeaders, options?: Tx.Options): Promise<Tx.Info>
+  async transaction(action: 'commit', headers: Tx.TransactionHeaders, info: Tx.Info): Promise<undefined>
+  async transaction(action: 'rollback', headers: Tx.TransactionHeaders, info: Tx.Info): Promise<undefined>
+  async transaction(action: any, headers: Tx.TransactionHeaders, arg?: any) {
+    await this.start()
+
+    const errorHandler = async (response: RequestResponse) => {
+      if (!response.ok) {
+        const json = await response.json()
+
+        if (json) {
+          throw new Error(`Error in transaction start: ${JSON.stringify(json)}`)
+        } else {
+          throw new Error(`Error in transaction start: ${response.text()}`)
+        }
+      }
+    }
+
+    if (action === 'start') {
+      const body = JSON.stringify({
+        max_wait: arg?.maxWait ?? 2000, // default
+        timeout: arg?.timeout ?? 5000, // default
+        isolation_level: arg?.isolationLevel,
+      })
+
+      const url = await this.url('transaction/start')
+
+      const response = await request(url, {
+        method: 'POST',
+        headers: { ...headers, ...this.headers },
+        body,
+        clientVersion: this.clientVersion,
+      })
+
+      await errorHandler(response)
+
+      const json = await response.json()
+
+      const endpoint = json['data-proxy'].endpoint as string
+
+      const id = endpoint.split('/').pop() as string
+
+      const data: Tx.Info = {
+        id,
+      }
+
+      return data as Tx.Info
```

Comment:
This ignores the hostname, which will be different for the transaction compared to the main hostname. This will work with mini-proxy right now but will not work with the real Data Proxy.